### ISSUE TITLE
Doesn't work in IE11?

### ISSUE THREAD

**Issue by jakearchibald** [filamentgroup/loadCSS#36]
I'm using `loadCSS` on [SVGOMG](https://jakearchibald.github.io/svgomg/). However, [it doesn't appear to work in IE11](http://www.webpagetest.org/video/compare.php?tests=150215_PF_G69-r:1-c:0) - initial render is blocked until CSS downloads.

I tried adding `ss.lazyLoad = 1`, but no change.

In the end I moved the `loadCSS` call after my content https://github.com/jakearchibald/svgomg/commit/d966d837a96f3d348b07e222302bd8ea45324177, and [that did the trick](http://www.webpagetest.org/video/compare.php?tests=150215_7S_JEP-r%3A1-c%3A0&thumbSize=200&ival=100&end=visual)

**Comment by scottjehl** [comment]
Thanks @jakearchibald 

Well... that's not good. I'll give this a look today. Maybe we could try other things like adding `rel=alternate stylesheet` or even... prefetch? Maybe not. :) 

Anyway... time for some tinkering I think.

**Comment by scottjehl** [comment]
Hey again, @jakearchibald 
Okay, quick update. I'm having no luck reproducing this issue, which of course is encouraging but I'm wondering if I might just be doing it wrong. :)

Here's what I did to test it:

Step 1. I created a CSS file that takes 5 seconds to respond and dropped it on my server. It's here http://scottjehl.com/css-temp/slow.php and its source reads:
``` php
<?php
sleep(5);
header("Content-Type: text/css");
echo "body { background: green; color: #fff; }";
?>
```
Step 2. I created a [control.html page](http://filamentgroup.github.io/loadCSS/control.html) to verify that an ordinary link to that CSS file will block render as expected. The stylesheet will style the background green and text white when it loads. Verified that it works as expected in IE11:
![control](https://cloud.githubusercontent.com/assets/214783/6218617/88b8ab8c-b5e5-11e4-8895-25854925603d.gif)

Step 3. I created a [test.html page](http://filamentgroup.github.io/loadCSS/test.html) and loaded the same stylesheet using loadCSS instead. My expectation is to see unstyled text immediately, and white-on-green whenever the stylesheet loads. Verified that it works as expected in IE11:
![test](https://cloud.githubusercontent.com/assets/214783/6218641/d6dd9c5a-b5e5-11e4-8dd5-ce1d88a66d84.gif)

Any tips on how to reproduce the issue would be great.

**Comment by jakearchibald** [comment]
Hm, I wonder what else was at play. I saw it on webpagetest but also a VM,
so I don't think it was a quirk in one system. Will try and reduce.

**Comment by scottjehl** [comment]
Thanks Jake!

**Comment by matijs** [comment]
I created a PR (https://github.com/filamentgroup/loadCSS/pull/37) that I tested on IE11. Worked as expected in IE11 in a (modern.ie) VM. Instead of polling, it relies on the load event of the stylesheet. Curious if that makes a difference.

**Comment by scottjehl** [comment]
We haven’t been able to reproduce a bug showing that the version in master has problems running in IE 11. 

My last message shows that it does appear to function as expected there, so I was leaving this open to see if @jakearchibald found a way to reproduce the issue.

**Comment by jakearchibald** [comment]
Yeah, sorry I haven't done this yet, will hopefully get to it tomorrow

**Comment by ausi** [comment]
I had the same issue and was able to reproduce it in a small test case: <http://madeyourday.net/load-css-test/load-css.html>

It’s basically the same setup as @scottjehl’s test, but I added a `<link>` to a fast stylesheet after the loadCSS-script:

```html
<head>
	<script>
		function loadCSS(){...}
		loadCSS('slow-style.php');
	</script>
	<link rel="stylesheet" href="fast-style.php">
</head>
<body>
	<p>Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam.</p>
</body>
```

I tested it with version 0.1.6 and the one from the master branch, both have the same issue.

**Comment by ausi** [comment]
I reduced it a little bit more by replacing the fast `<link>` with an empty script tag: <http://madeyourday.net/load-css-test/load-css-2.html>

```html
<head>
	<script>
		function loadCSS(){...}
		loadCSS('slow-style.php');
	</script>
	<script></script>
</head>
<body>
	<p>Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam.</p>
</body>
```

This blocks the rendering of the whole page in IE11 until *slow-style.php* is loaded. If I remove the second `<script>` tag it doesn’t block.

**Comment by melnikovkolya** [comment]
Works for me on Win 8...

**Comment by ausi** [comment]
It still doesn’t work in my IE downloaded from modern.ie (Version 11.0.9600.17031 on Windows 8.1).

@melnikovkolya Did you see black text on a white background for about five seconds before it switched to white text on a green background?

**Comment by melnikovkolya** [comment]
I created this plunk [http://plnkr.co/edit/LjksZH0CWtQ8huQ0kM1B?p=preview] and it seems to work on the native Windows 8, IE 11

**Comment by ausi** [comment]
@melnikovkolya Your code is completely different, please take a look at the example i posted in https://github.com/filamentgroup/loadCSS/issues/36#issuecomment-89058404 or just click the link [madeyourday.net/load-css-test/load-css-2.html](http://madeyourday.net/load-css-test/load-css-2.html) and check if you can see black text before the background is green.

**Comment by melnikovkolya** [comment]
My bad for not realizing the real issue. I just tried the link you provided and I do not notice any black text - the page loaded.

**Comment by ausi** [comment]
> I do not notice any black text

Exactly that is the problem this issue is all about. Take a look at the animated GIFs in https://github.com/filamentgroup/loadCSS/issues/36#issuecomment-74566021

**Comment by melnikovkolya** [comment]
I just tried your example on plunker and it Works as expected, e.g.: text in black is loaded first, once the timeout expires ' a new style is applied: [http://plnkr.co/edit/cBOLboPLIiK7jNek1iat?p=preview]. The only thing I changed from your original code is: shifted the loadCSS call into the body.

The weird part is: I reran/reloaded this example a couple of times and not it doesn´t behave the same way any more. I tried Ctrl+F5, still not the same as the init behaviour (the expected one)... this is weird

**Comment by melnikovkolya** [comment]
now there is even no 5 sec blocking

**Comment by ausi** [comment]
> The only thing I changed from your original code is: shifted the loadCSS call into the body.

This is the workaround @jakearchibald pointed out in https://github.com/filamentgroup/loadCSS/issues/36#issue-57735160, but this issue isn’t about a workaround but to find out why IE11 does block rendering in certain cases.

> The weird part is: I reran/reloaded this example a couple of times and not it doesn´t behave the same way any more. I tried Ctrl+F5, still not the same as the init behaviour (the expected one)... this is weird

It isn’t easy to debug, try the following:

1. open a new empty tab
2. clear the browser cache
3. open the URL

**Comment by scottjehl** [comment]
I just tested this again and can't seem to reproduce a problem in IE11 with the latest loadCSS. If anyone is able to create a demo that reproduces the issue, we're more than happy to check it out. Thanks!

**Comment by ausi** [comment]
@scottjehl I updated my test case to the latest version of loadCSS: [madeyourday.net/load-css-test/load-css-3.html](http://madeyourday.net/load-css-test/load-css-3.html) and it is still blocking rendering in IE11.

![ie11-loadcss](https://cloud.githubusercontent.com/assets/367169/9549935/d1be7b08-4da8-11e5-8bf1-277fd3a46997.gif)

**Comment by scottjehl** [comment]
Thanks @ausi

**Comment by scottjehl** [comment]
Hi @ausi 

Does that empty script tag in your example have any bearing on whether the request blocks in IE11?

**Comment by ausi** [comment]
@scottjehl Yes, if you remove the empty `<script>` tag it doesn’t block.

**Comment by scottjehl** [comment]
@ausi interesting. Hey @jakearchibald is that similar to your example perhaps? Maybe there's a `script` element after the one containing loadCSS in `head`?

**Comment by scottjehl** [comment]
Hmm... after a quick check, looks like svgomg might just be waiting for `request animation frame` before requesting CSS now, so loadCSS isn't in play there.

**Comment by ausi** [comment]
As you can see in my first example (https://github.com/filamentgroup/loadCSS/issues/36#issuecomment-88412666) it also blocks if you use a fast stylesheet via `<link>` after the loadCSS script. So it seems that if there is a `<script>` or `<link>` in the `<head>` after the loadCSS script, the rendering is blocked in IE11.

**Comment by scottjehl** [comment]
@ausi thanks. Would you be able to update your loadCSS and see if the issue persists? I'm curious if the new default insertion point (which is *after* the last style/link/script in the DOM) affects this bug?

**Comment by ausi** [comment]
@scottjehl I still get the same results with the newest version from the master branch:

* [load-css-4.html](http://madeyourday.net/load-css-test/load-css-4.html): loadCSS **with** empty `<script>` tag following it **blocks** rendering in IE11
* [load-css-5.html](http://madeyourday.net/load-css-test/load-css-5.html): loadCSS **without** `<script>` tag following it **doesn’t block**  rendering in IE11

**Comment by scottjehl** [comment]
@ausi thanks!

**Comment by DmitriyTrt** [comment]
Just to clarify, processing and rendering of the document freezes on any <script> or <link rel="stylesheet"> found in the document after the script calling loadCSS() on IE 11. It has to be a separate <script> tag to trigger it, code on the same <script> tag executes before the freeze. So moving code to the body is only a partial solution, unless it's really in the end of all body scripts.

Tested on IE v11.0.8 Windows Server 2012 R2, sorry don't have anything more fresh at the moment.

**Comment by scottjehl** [comment]
Back to prioritizing this one now that the Chrome issue and polyfill is resolved.

We should test the latest script now in master, and probably the pattern using a polyfill of rel=preload as well:

http://filamentgroup.github.io/loadCSS/test/preload.html

**Comment by scottjehl** [comment]
I did some testing on this today. Inserting an empty (or populated) script tag after the one that contains loadCSS will cause a render-blocking CSS load in IE11, as well as Edge, IE10, and probably older IEs as well but I didn't test.

So! Two approaches come to mind:

A. We could just add a recommendation to the README saying that the script tag that calls loadCSS should be at the end of the `head`, with no scripts or external stylesheets following it in the source.
B. We could add a workaround to make sure it works anyway.

If we want to go with B, I've got a workaround ready and it tests out fine in Edge, IE11, and IE10. The change involves waiting until the `body` element is defined before inserting the CSS link. This means a request could possible go out milliseconds later than it otherwise would, but that's likely not a problem considering this script is used to purposely allow a page to render while lower-priority css loads.

PR coming for approach B...
